File: rocket-bi-server/src/main/scala/co/datainsider/jobscheduler/domain/source/MixpanelSource.scala
package co.datainsider.jobscheduler.domain.source

import co.datainsider.jobscheduler.domain.Ids.SourceId
import co.datainsider.jobscheduler.domain.source.DataSourceType.DataSourceType
import co.datainsider.jobscheduler.domain.source.MixpanelRegion.MixpanelRegion
import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
case class MixpanelSource(
    orgId: Long,
    id: SourceId,
    displayName: String,
    accountUsername: String,
    accountSecret: String,
    projectId: String,
    @JsonScalaEnumeration(classOf[MixpanelRegionRef])
    region: MixpanelRegion = MixpanelRegion.US,
    timezone: String = "US/Pacific",
    creatorId: String = "",
    lastModify: Long = System.currentTimeMillis()
) extends DataSource {
  override def getId: SourceId = id

  override def getName: String = displayName

  override def getType: DataSourceType = DataSourceType.Mixpanel

  override def getConfig: Map[String, Any] =
    Map(
      "account_username" -> accountUsername,
      "account_secret" -> accountSecret,
      "project_id" -> projectId,
      "region" -> region.toString,
      "timezone" -> timezone
    )

  override def getCreatorId: String = creatorId

  override def getLastModify: Long = lastModify
}

object MixpanelRegion extends Enumeration {
  type MixpanelRegion = Value
  val US = Value("US")
  val EU = Value("EU")
}

class MixpanelRegionRef extends TypeReference[MixpanelRegion.type]

File: web/di-core/data-ingestion/domain/response/MixpanelSource.ts
import { DataSources, MixpanelRegion } from '@core/data-ingestion';
import { DataSource } from '@core/data-ingestion/domain/response/DataSource';
import { SourceId } from '@core/common/domain';

export class MixpanelSource implements DataSource {
  readonly className = DataSources.Mixpanel;
  id: SourceId;
  orgId: string;
  displayName: string;
  accountUsername: string;
  accountSecret: string;
  projectId: string;
  region: MixpanelRegion;
  timezone: string;
  lastModify: number;

  constructor(
    id: SourceId,
    orgId: string,
    displayName: string,
    accountUsername: string,
    accountSecret: string,
    projectId: string,
    region: MixpanelRegion,
    timezone: string,
    lastModify: number
  ) {
    this.id = id;
    this.orgId = orgId;
    this.displayName = displayName;
    this.lastModify = lastModify;
    this.accountUsername = accountUsername;
    this.accountSecret = accountSecret;
    this.projectId = projectId;
    this.region = region;
    this.timezone = timezone;
  }

  static fromObject(obj: any): MixpanelSource {
    return new MixpanelSource(
      obj.id,
      obj.orgId,
      obj.displayName,
      obj.accountUsername,
      obj.accountSecret,
      obj.projectId,
      obj.region,
      obj.timezone,
      obj.lastModify
    );
  }
}
